File: backend-java/src/main/java/com/streaming/features/video/PresignService.java
package com.streaming.features.video;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Map;

@Service
public class PresignService {

    @Value("${APP_URL:http://localhost:3000}")
    private String appUrl;

    @Value("${PRESIGN_SECRET:mysecret}")
    private String presignSecret;

    private final VideoService videoService;

    public PresignService(VideoService videoService) {
        this.videoService = videoService;
    }

    public Map<String, Object> generateSignedUrl(String id) {
        String streamId = videoService.getPublicStreamId(id);
        long expiresAt = Instant.now().getEpochSecond() + 60;
        String token = sign(streamId + ":" + expiresAt);

        return Map.of(
                "url", String.format("%s/video/stream/%s?token=%s&exp=%d", appUrl, streamId, token, expiresAt),
                "expiresAt", expiresAt,
                "streamId", streamId
        );
    }

    public void validateSignature(String id, long exp, String token) {
        long now = Instant.now().getEpochSecond();
        if (now > exp) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Presigned URL expired");
        }

        String expected = sign(id + ":" + exp);

        if (!java.security.MessageDigest.isEqual(expected.getBytes(StandardCharsets.UTF_8), token.getBytes(StandardCharsets.UTF_8))) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid token");
        }
    }

    private String sign(String payload) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKeySpec = new SecretKeySpec(presignSecret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            mac.init(secretKeySpec);
            byte[] hash = mac.doFinal(payload.getBytes(StandardCharsets.UTF_8));

            // Convert to hex
            StringBuilder hexString = new StringBuilder(2 * hash.length);
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) {
                    hexString.append('0');
                }
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (Exception e) {
            throw new RuntimeException("Failed to calculate HMAC", e);
        }
    }
}

File: backend-nest/src/video/presign.service.ts
import { ForbiddenException, Injectable, UnauthorizedException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as crypto from 'crypto';
import { VideoService } from './video.service';

@Injectable()
export class PresignService {
  constructor(
    private readonly configService: ConfigService,
    private readonly videoService: VideoService,
  ) {}

  generateSignedUrl(id: string) {
    const streamId = this.videoService.getPublicStreamId(id);
    const expiresAt = Math.floor(Date.now() / 1000) + 60;
    const token = this.sign(`${streamId}:${expiresAt}`);
    const baseUrl = this.configService.get<string>('APP_URL') || 'http://localhost:3000';

    return {
      url: `${baseUrl}/video/stream/${streamId}?token=${token}&exp=${expiresAt}`,
      expiresAt,
      streamId,
    };
  }

  validateSignature(id: string, exp: number, token: string) {
    const now = Math.floor(Date.now() / 1000);
    if (now > exp) {
      throw new ForbiddenException('Presigned URL expired');
    }

    const expected = this.sign(`${id}:${exp}`);
    const provided = Buffer.from(token, 'hex');
    const actual = Buffer.from(expected, 'hex');

    if (provided.length !== actual.length || !crypto.timingSafeEqual(provided, actual)) {
      throw new UnauthorizedException('Invalid token');
    }
  }

  private sign(payload: string) {
    const secret = this.configService.get<string>('PRESIGN_SECRET') || 'mysecret';
    return crypto.createHmac('sha256', secret).update(payload).digest('hex');
  }
}
